File: app/src/main/java/org/secuso/privacyfriendlyboardgameclock/activities/PlayerManagementActivity.java
/*
 This file is part of Privacy Friendly Board Game Clock.

 Privacy Friendly Board Game Clock is free software:
 you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation,
 either version 3 of the License, or any later version.

 Privacy Friendly Board Game Clock is distributed in the hope
 that it will be useful, but WITHOUT ANY WARRANTY; without even
 the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Privacy Friendly Board Game Clock. If not, see <http://www.gnu.org/licenses/>.
 */

package org.secuso.privacyfriendlyboardgameclock.activities;

import android.app.AlertDialog;
import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;

import org.secuso.privacyfriendlyboardgameclock.R;
import org.secuso.privacyfriendlyboardgameclock.database.PlayersDataSourceSingleton;
import org.secuso.privacyfriendlyboardgameclock.fragments.PlayerManagementChooseModeFragment;
import org.secuso.privacyfriendlyboardgameclock.fragments.PlayerManagementContactListFragment;
import org.secuso.privacyfriendlyboardgameclock.fragments.PlayerManagementEditPlayerFragment;
import org.secuso.privacyfriendlyboardgameclock.helpers.ItemClickListener;
import org.secuso.privacyfriendlyboardgameclock.helpers.PlayerListAdapter;
import org.secuso.privacyfriendlyboardgameclock.helpers.TAGHelper;
import org.secuso.privacyfriendlyboardgameclock.model.Player;

import java.util.List;

/**
 * Actionbar Tutorial
 * http://www.vogella.com/tutorials/AndroidActionBar/article.html#exercise-using-the-contextual-action-mode
 * Selection State tutorial
 * https://enoent.fr/blog/2015/01/18/recyclerview-basics/
 * Created by Quang Anh Dang on 24.12.2017.
 * @author Quang Anh Dang
 * Last changed on 18.03.18
 * This is the Activity for the Player Management Page
 */
public class PlayerManagementActivity extends BaseActivity implements ItemClickListener{
    @SuppressWarnings("unused")
    private static final String TAG = MainActivity.class.getSimpleName();

    private PlayersDataSourceSingleton pds;
    private RecyclerView playersRecycleView;
    private PlayerListAdapter playerListAdapter;
    private List<Player> listPlayers;
    private LinearLayoutManager layoutManager;
    private FloatingActionButton fabAdd;
    private FloatingActionButton fabDelete;
    private int selectedPlayerId = -1;
    private FragmentManager fm;
    // To toggle selection mode
    private ActionModeCallback actionModeCallback = new ActionModeCallback();
    private ActionMode actionMode;
    private Player playerToEdit;
    private View insertAlert;
    private View emptyListLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_player_management);
        fm = getFragmentManager();
        // pds already opened in MainActivity
        pds = PlayersDataSourceSingleton.getInstance(getApplicationContext());

        listPlayers = pds.getAllPlayers();
        layoutManager = new LinearLayoutManager(this);

        // FAB Listener
        fabAdd = findViewById(R.id.fab_add_new_player);
        fabAdd.setOnClickListener(onFABAddClickListener());
        fabDelete = findViewById(R.id.fab_delete_player);
        fabDelete.setOnClickListener(onFABDeleteListenter());

        insertAlert = findViewById(R.id.insert_alert);
        emptyListLayout = findViewById(R.id.emptyListLayout);

        playersRecycleView = findViewById(R.id.player_list);
        playersRecycleView.setHasFixedSize(true);
        playerListAdapter = new PlayerListAdapter(this, listPlayers, this);
        playersRecycleView.setAdapter(playerListAdapter);
        playersRecycleView.setLayoutManager(layoutManager);
        playersRecycleView.setItemAnimator(null);

        Animation anim = new AlphaAnimation(0.0f, 1.0f);
        anim.setDuration(1500); //You can manage the blinking time with this parameter
        anim.setStartOffset(20);
        anim.setRepeatMode(Animation.REVERSE);
        anim.setRepeatCount(Animation.INFINITE);
        insertAlert.startAnimation(anim);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode){
            case TAGHelper.REQUEST_READ_CONTACT_CODE:{
                // If request is cancelled, the result arrays are empty.
                if(grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
                    // Permission granted
                    FragmentManager fm = getFragmentManager();
                    FragmentTransaction ft = fm.beginTransaction();
                    Fragment prev = fm.findFragmentByTag("dialog");
                    if(prev != null) ft.remove(prev);
                    ft.addToBackStack(null);

                    // Create and show the dialog
                    PlayerManagementContactListFragment contactListFragment = new PlayerManagementContactListFragment();
                    contactListFragment.show(ft, "dialog");
                }
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if(playerListAdapter.getPlayersList().size() == 0){
            insertAlert.setVisibility(View.VISIBLE);
            emptyListLayout.setVisibility(View.VISIBLE);
        } else {
            insertAlert.setVisibility(View.GONE);
            emptyListLayout.setVisibility(View.GONE);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onItemClick(View view, final int position) {
        if(actionMode != null){
            toggleSelection(position);
        } else{
            FragmentTransaction ft = fm.beginTransaction();
            Fragment prev = fm.findFragmentByTag(TAGHelper.DIALOG_FRAGMENT);
            if(prev != null) ft.remove(prev);
            ft.addToBackStack(null);
            playerToEdit = playerListAdapter.getPlayer(position);

            // Create and show the dialog
            PlayerManagementEditPlayerFragment editPlayerFragment = PlayerManagementEditPlayerFragment.newInstance("Edit Player");
            editPlayerFragment.show(ft,TAGHelper.DIALOG_FRAGMENT);
        }
    }

    @Override
    public boolean onItemLongClicked(View view, int position) {
        if (actionMode == null) {
            actionMode = startSupportActionMode(actionModeCallback);
        }
        toggleSelection(position);
        return true;
    }

    /**
     * Toggle the selection state of an item.
     *
     * If the item was the last one in the selection and is unselected, the selection is stopped.
     * Note that the selection must already be started (actionMode must not be null).
     *
     * @param position Position of the item to toggle the selection state
     */
    private void toggleSelection(int position) {
        playerListAdapter.toggleSelection(position);
        int count = playerListAdapter.getSelectedItemCount();

        if (count == 0) {
            actionMode.finish();
        } else {
            actionMode.setTitle(String.valueOf(count));
            actionMode.invalidate();
        }
    }

    @Override
    protected int getNavigationDrawerID() {return R.id.nav_player_management;}

    /**
     * open Dialogfragment for user to choose how to create new player
     * @return
     */
    private View.OnClickListener onFABAddClickListener(){
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                FragmentTransaction ft = fm.beginTransaction();
                Fragment prev = getFragmentManager().findFragmentByTag(TAGHelper.DIALOG_FRAGMENT);
                if(prev != null) ft.remove(prev);
                ft.addToBackStack(null);

                // Create and show the dialog
                PlayerManagementChooseModeFragment chooseDialogFragment = PlayerManagementChooseModeFragment.newInstance("Choose how to create new player:");
                chooseDialogFragment.show(ft,TAGHelper.DIALOG_FRAGMENT);
            }
        };
    }

    /**
     * remove all the selected players
     * @return
     */
    private View.OnClickListener onFABDeleteListenter() {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new AlertDialog.Builder(PlayerManagementActivity.this)
                        .setTitle(R.string.warning)
                        .setMessage(R.string.playerDeleteWarning)
                        .setIcon(android.R.drawable.ic_dialog_alert)
                        .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                playerListAdapter.removeItems(playerListAdapter.getSelectedItems());
                                actionMode.finish();
                                actionMode = null;
                                if(playerListAdapter.getPlayersList().size() == 0){
                                    insertAlert.setVisibility(View.VISIBLE);
                                    emptyListLayout.setVisibility(View.VISIBLE);
                                } else {
                                    insertAlert.setVisibility(View.GONE);
                                    emptyListLayout.setVisibility(View.GONE);
                                }
                            }
                        })
                        .setNegativeButton(R.string.no, null)
                        .show();
            }
        };
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (!drawer.isDrawerOpen(GravityCompat.START) && fm.getBackStackEntryCount() > 0) {
            finish();
            startActivity(getIntent());
        }
        else super.onBackPressed();
    }

    /**
     * Infalte the Actionicons on Toolbar, in this case the delete icon
     * @param menu
     * @return
     */
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.player_management_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.action_delete:
                if(actionMode == null)
                    actionMode = startSupportActionMode(actionModeCallback);
        }
        return true;
    }

    private void switchVisibilityOf2FABs(){
        if(fabAdd.getVisibility() != View.GONE && fabDelete.getVisibility() == View.GONE){
            fabAdd.setVisibility(View.GONE);
            fabDelete.setVisibility(View.VISIBLE);
        }else{
            fabAdd.setVisibility(View.VISIBLE);
            fabDelete.setVisibility(View.GONE);
        }
    }

    public Player getPlayerToEdit() {
        return playerToEdit;
    }

    public void setPlayerToEdit(Player playerToEdit) {
        this.playerToEdit = playerToEdit;
    }

    /*  #################################################################
        #                                                               #
        #                       Helper class                            #
        #                                                               #
        #################################################################*/

    private class ActionModeCallback implements ActionMode.Callback {
        @SuppressWarnings("unused")
        private final String TAG = ActionModeCallback.class.getSimpleName();

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            playerListAdapter.setLongClickedSelected(true);
            playerListAdapter.setSimpleClickedSelected(false);
            mode.getMenuInflater().inflate (R.menu.selected_menu, menu);
            switchVisibilityOf2FABs();
            // so all check box are visible
            playerListAdapter.clearSelection();
            playerListAdapter.notifyDataSetChanged();
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            playerListAdapter.setLongClickedSelected(false);
            playerListAdapter.clearSelection();
            actionMode = null;
            switchVisibilityOf2FABs();
            // so all check box are visible
            playerListAdapter.clearSelection();
            playerListAdapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/org/secuso/privacyfriendlyboardgameclock/activities/PlayerManagementActivity.kt
/*
 This file is part of Privacy Friendly Board Game Clock.

 Privacy Friendly Board Game Clock is free software:
 you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation,
 either version 3 of the License, or any later version.

 Privacy Friendly Board Game Clock is distributed in the hope
 that it will be useful, but WITHOUT ANY WARRANTY; without even
 the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Privacy Friendly Board Game Clock. If not, see <http://www.gnu.org/licenses/>.
 */
package org.secuso.privacyfriendlyboardgameclock.activities

import android.database.Cursor
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import androidx.appcompat.view.ActionMode
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import org.secuso.pfacore.model.DrawerElement
import org.secuso.privacyfriendlyboardgameclock.R
import org.secuso.privacyfriendlyboardgameclock.helpers.ItemClickListener
import org.secuso.privacyfriendlyboardgameclock.helpers.PlayerListAdapter
import org.secuso.privacyfriendlyboardgameclock.room.model.Player
import androidx.core.view.isGone
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import kotlinx.coroutines.launch
import org.secuso.pfacore.model.dialog.AbortElseDialog
import org.secuso.pfacore.ui.dialog.show
import org.secuso.privacyfriendlyboardgameclock.activities.game.buildChooseNewPlayerCreationMethodDialog
import org.secuso.privacyfriendlyboardgameclock.activities.game.buildCreateNewPlayerDialog
import org.secuso.privacyfriendlyboardgameclock.activities.game.buildCreatePlayerFromContactDialog
import org.secuso.privacyfriendlyboardgameclock.activities.game.useCameraForPlayerPicture
import org.secuso.privacyfriendlyboardgameclock.activities.game.useContactsForAddingPlayers

/**
 * Actionbar Tutorial
 * http://www.vogella.com/tutorials/AndroidActionBar/article.html#exercise-using-the-contextual-action-mode
 * Selection State tutorial
 * https://enoent.fr/blog/2015/01/18/recyclerview-basics/
 * Created by Quang Anh Dang on 24.12.2017.
 * @author Quang Anh Dang
 * Last changed on 18.03.18
 * This is the Activity for the Player Management Page
 */
class PlayerManagementActivity : BaseActivity(), ItemClickListener {
    private val viewModel by lazy { ViewModelProvider(this)[PlayerManagementActivityViewModel::class.java] }
    private val playersRecycleView by lazy { findViewById<RecyclerView>(R.id.player_list) }
    private lateinit var playerListAdapter: PlayerListAdapter
    private var listPlayers: MutableList<Player> = mutableListOf()
    private val fabAdd by lazy { findViewById<FloatingActionButton>(R.id.fab_add_new_player) }
    private val fabDelete by lazy { findViewById<FloatingActionButton>(R.id.fab_delete_player) }

    // To toggle selection mode
    private val actionModeCallback: ActionModeCallback = ActionModeCallback()
    private var actionMode: ActionMode? = null
    @JvmField
    var playerToEdit: Player? = null
    var playerToEditIndex: Int? = null
    private val insertAlert by lazy { findViewById<View>(R.id.insert_alert) }
    private val emptyListLayout by lazy { findViewById<View>(R.id.emptyListLayout) }

    override fun isActiveDrawerElement(element: DrawerElement) = element.icon == R.drawable.ic_menu_player_management

    private lateinit var pictureConsumer: (Bitmap) -> Unit
    private val useCamera = useCameraForPlayerPicture {
        val bitmap = it.planes[0].buffer.let { buffer ->
            val bytes = ByteArray(buffer.capacity())
            buffer[bytes]
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size, null)
        }
        pictureConsumer(bitmap)
    }

    private val createNewPlayerDialog by lazy {
        buildCreateNewPlayerDialog(
            onPlayerCreated = {
                viewModel.addPlayer(it)
                playerListAdapter.playersList.add(it)
                playerListAdapter.notifyItemInserted(playerListAdapter.itemCount - 1)
            },
            useCamera = {
                pictureConsumer = it
                useCamera()
            }
        )
    }

    private var data: Cursor? = null
    private val createNewPlayerFromContactDialog by lazy {
        buildCreatePlayerFromContactDialog({ data }) { name, icon ->
            viewModel.addPlayer(name, icon)
        }
    }
    private val useContacts = useContactsForAddingPlayers {
        data = it
        createNewPlayerFromContactDialog.show()
    }

    private val choosePlayerCreationMethod by lazy {
        buildChooseNewPlayerCreationMethodDialog(
            useContacts,
            createNewPlayerDialog
        )
    }

    private val editPlayerDialog by lazy {
        buildCreateNewPlayerDialog(
            onPlayerCreated = {
                viewModel.updatePlayer(it)
                playerListAdapter.playersList[playerToEditIndex!!] = it
                playerListAdapter.notifyItemChanged(playerToEditIndex!!)
                playerToEditIndex = null
                playerToEdit = null
            },
            useCamera = {
                pictureConsumer = it
                useCamera()
            },
            playerSupplier = { playerToEdit!! }
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_player_management)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.CREATED) {
                viewModel.getAllPlayers().collect {
                    playerListAdapter.playersList = it
                    playerListAdapter.notifyDataSetChanged()
                }
            }
        }
        listPlayers = viewModel.getAllPlayersSync().toMutableList()

        // FAB Listener
        fabAdd.setOnClickListener {
            choosePlayerCreationMethod.show()
        }
        fabDelete.setOnClickListener {
            AbortElseDialog.build(this) {
                title = { ContextCompat.getString(this@PlayerManagementActivity, R.string.warning) }
                content = { ContextCompat.getString(this@PlayerManagementActivity, R.string.playerDeleteWarning) }
                icon = android.R.drawable.ic_dialog_alert
                acceptLabel = ContextCompat.getString(this@PlayerManagementActivity, R.string.yes)
                abortLabel = ContextCompat.getString(this@PlayerManagementActivity, R.string.no)

                onElse = {
                    lifecycleScope.launch {
                        playerListAdapter.selectedItems
                            .map { playerListAdapter.playersList[it] }
                            .forEach { viewModel.removePlayer(it) }
                    }
                    playerListAdapter.removeItems(playerListAdapter.getSelectedItems())
                    actionMode?.finish()
                    actionMode = null
                    if (playerListAdapter.playersList.isEmpty()) {
                        insertAlert.visibility = View.VISIBLE
                        emptyListLayout.visibility = View.VISIBLE
                    } else {
                        insertAlert.visibility = View.GONE
                        emptyListLayout.visibility = View.GONE
                    }
                }
            }.show()
        }

        playerListAdapter = PlayerListAdapter(this, listPlayers, this)
        playersRecycleView.apply {
            setHasFixedSize(true)
            adapter = playerListAdapter
            layoutManager = LinearLayoutManager(this@PlayerManagementActivity)
            itemAnimator = null
        }

        val anim: Animation = AlphaAnimation(0.0f, 1.0f).apply {
            duration = 1500
            startOffset = 20
            repeatMode = Animation.REVERSE
            repeatCount = Animation.INFINITE
        }
        insertAlert.startAnimation(anim)
    }

    override fun onResume() {
        super.onResume()
        if (playerListAdapter.playersList.isEmpty()) {
            insertAlert.visibility = View.VISIBLE
            emptyListLayout.visibility = View.VISIBLE
        } else {
            insertAlert.visibility = View.GONE
            emptyListLayout.visibility = View.GONE
        }
    }

    override fun onItemClick(view: View?, position: Int) {
        if (actionMode != null) {
            toggleSelection(position)
        } else {
            playerToEdit = playerListAdapter.getPlayer(position)
            playerToEditIndex = position
            editPlayerDialog.show()
        }
    }

    override fun onItemLongClicked(view: View?, position: Int): Boolean {
        if (actionMode == null) {
            actionMode = startSupportActionMode(actionModeCallback)
        }
        toggleSelection(position)
        return true
    }

    /**
     * Toggle the selection state of an item.
     *
     * If the item was the last one in the selection and is unselected, the selection is stopped.
     * Note that the selection must already be started (actionMode must not be null).
     *
     * @param position Position of the item to toggle the selection state
     */
    private fun toggleSelection(position: Int) {
        playerListAdapter.toggleSelection(position)
        val count = playerListAdapter.selectedItemCount

        if (count == 0) {
            actionMode?.finish()
        } else {
            actionMode?.title = count.toString()
            actionMode?.invalidate()
        }
    }

    /**
     * Infalte the Actionicons on Toolbar, in this case the delete icon
     * @param menu
     * @return
     */
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.player_management_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_delete) {
            if (actionMode == null) actionMode = startSupportActionMode(actionModeCallback)
        } else {
            return super.onOptionsItemSelected(item)
        }
        return true
    }

    private fun switchVisibilityOf2FABs() {
        if (fabAdd.visibility != View.GONE && fabDelete.isGone) {
            fabAdd.visibility = View.GONE
            fabDelete.visibility = View.VISIBLE
        } else {
            fabAdd.visibility = View.VISIBLE
            fabDelete.visibility = View.GONE
        }
    }

    /*  #################################################################
         #                                                               #
         #                       Helper class                            #
         #                                                               #
         #################################################################*/
    private inner class ActionModeCallback : ActionMode.Callback {
        @Suppress("unused")
        private val TAG: String = ActionModeCallback::class.java.simpleName

        override fun onCreateActionMode(mode: ActionMode, menu: Menu?): Boolean {
            playerListAdapter.setLongClickedSelected(true)
            playerListAdapter.setSimpleClickedSelected(false)
            mode.menuInflater.inflate(R.menu.selected_menu, menu)
            switchVisibilityOf2FABs()
            // so all check box are visible
            playerListAdapter.clearSelection()
            playerListAdapter.notifyDataSetChanged()
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
            return false
        }

        override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean {
            return false
        }

        override fun onDestroyActionMode(mode: ActionMode?) {
            playerListAdapter.setLongClickedSelected(false)
            playerListAdapter.clearSelection()
            actionMode = null
            switchVisibilityOf2FABs()
            // so all check box are visible
            playerListAdapter.clearSelection()
            playerListAdapter.notifyDataSetChanged()
        }
    }

    companion object {
        @Suppress("unused")
        private val TAG: String = MainActivity::class.java.simpleName
    }
}